starknet_py/cairo/type_parser.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import annotations
+from collections import OrderedDict
+from typing import Dict
+
+from starkware.cairo.lang.compiler.parser import parse_type
+import starkware.cairo.lang.compiler.ast.cairo_types as cairo_lang_types
+
+from starknet_py.cairo.data_types import (
+    CairoType,
+    FeltType,
+    TupleType,
+    NamedTupleType,
+    ArrayType,
+    StructType,
+)
+
+
+class UnknownTypeError(ValueError):
+    """
+    Error thrown when TypeParser finds type that was not declared prior to parsing.
+    """
+
+    type_name: str
+
+    def __init__(self, type_name: str):
+        super().__init__(f"Type '{type_name}' is not defined")
+        self.type_name = type_name
+
+
+class TypeParser:
+    """
+    Low level utility class for parsing Cairo types that can be used in external methods.
+    """
+
+    defined_types: Dict[str, StructType]
+
+    def __init__(self, defined_types: Dict[str, StructType]):
+        """
+        TypeParser constructor.
+
+        :param defined_types: dictionary containing all defined types. For now they can only be structures.
+        """
+        self.defined_types = defined_types
+        for name, struct in defined_types.items():
+            if name != struct.name:
+                raise ValueError(
+                    f"Keys must match name of type, [{name}] != [{struct.name}]"
```

Comment:
Should this error message wrap the types in square brackets `[]`?
It might seem like a `list` at first glance.
`UnknownTypeError` has the type wrapped in single quotes - `f"Type '{type_name}' is not defined"`.